docs/LOGGING.md

```diff
@@ -30,28 +30,49 @@ Logs coming from xen-tools container are of three kinds:
+
+Once the gzip log files are uploaded to the cloud, any log entries in them will not be available on the device. For any log files are still waiting to be uploaded, they are in the '/persist/newlog/devUpload' and '/persist/newlog/appUpload' directories. If a direct ssh connection to the device is possible, the log entries can be looked at in those directories by using "zcat" on gzip-filename or using "*".
 
 ## Log export to cloud
 
-Rsyslogd is configured to send logs via TCP socket to logmanager. Logmanager bundles the logs into protobuf messages and exports them to cloud using API. Today when a device loses network connectivity to cloud, there is a back-off mechanism built into the logmanager that prompts rsyslogd to stop sending logs. When there is network drop, logmanager stops reading log messages from rsyslogd (TCP socket). Rsyslogd stops sending logs when it's output socket buffer is full. As a result following logs will be queued on disk. After the device's network connectivity is restored, rsyslogd starts sending logs from it's disk queues.
+"loguploader" is a pillar service which is responsible for upload the gzip log files to the cloud server. The binary data of a gzip file is the payload portion of the authentication protobuf envolope structure. This is similar to all the other EVE POST messages, except that in those messsages the payload usually is data of another protobuf structure.
+
+The upload is one gzip file at a time. The "loguploader" finds the earliest timestamp from the gzip file's filename and sends the data to the cloud server. If the upload is sucessful, then the uploaded gzip file is removed from the directory. If the upload encounters an error, it will come back to retry again. There can be several different failure cases:
+
+1) the upload has no reply from the server and is TCP timed out
+2) the upload gets http status code of 503
+3) the upload gets http status code of 4xx
+4) the upload gets other types of error
+
+For the case 1 and 2, "loguploader" will just come back to retry the same gzip file again (the earliest in timestamp). After it repeats this 3 times, the "uploader" will mark the cloud server as "unreachable". This "unreachable" status affects the disk space management by "newlogd" as mentioned above. For case 3, if it repeats continously for 10 times for the same gzip file, this gzip file will be moved into "/persist/newlog/failedUpload" directory, this is to prevent one bad file stops the other log files to be uploaded forever. There is no special action for case 4, the "loguploader" will come back to retry again later.
```

Comment:
Do we have counters to track the number of failures/retries and in particular the number of files/entried which have been moved to the failedUpload directory?
If so it makes sense to mention that in the document.

Reply:
Can failedUpload grow over time, or is there some garbage collection to ensure this directory has an upper size bound?

Ditto for all of newlog - is there some disk space limit and what happens when it runs low?
For example, in the future with ZFS we could set a limit on the size of the directory which means that the logging code would see errors when it is exceeded and would need a strategy to prioritize how to use the disk space (for collect vs. upload vs. failedUpload).

Reply:
yes, i have counters in a map indexed by filename string to track this. will add some text here.
the 'failedUpload' directory is capped by currently in the code by 1000 files, about 40M bytes of space.
for the 'newlog', the '/persist' partition space available is monitored every 30 minutes, if below say 100M, we'll start to recycle just like it's connection failed/disconnected.

Reply:
So having some separate section on how we limit the disk space usage would be good. THen you can put that info about 1000 files/40M for failedUpload in that section.

Reply:
ok.